### PR TITLE
Relocation improvements

### PR DESCRIPTION
**Pull request by ret2libc**
This PR does some things:
1. it does not add `Cd` metadata when the relocation is in an executable section
2. it fixes relocations that point to sections (before this patch indeed, sections were excluded by the symbol table)
3. it groups together relocs that have the same import function.

### CODE REVIEW

**Comment by ret2libc** [comment]
NOT READY YET.

**Comment by radare** [comment]
wtf
```
@@ -1,4 +1,23 @@
 [Symbols]
+001 0x00001000 0x00010000  LOCAL SECTION    0 
+002 0x000124f8 0x000214f8  LOCAL SECTION    0 
+003 0x00013620 0x00022620  LOCAL SECTION    0 
+004 0x00013628 0x00022628  LOCAL SECTION    0 
+005 0x00013630 0x00022630  LOCAL SECTION    0 
+006 0x00015060 0x00024060  LOCAL SECTION    0 
+007 0x00015070 0x00024070  LOCAL SECTION    0 
+008 0x00016070 0x00025070  LOCAL SECTION    0 
+009 0x00016168 0x00025168  LOCAL SECTION    0 
+010 0x00016198 0x00025198  LOCAL SECTION    0 
+011 0x00000000 0x00000000  LOCAL SECTION    0 
+012 0x00000000 0x00000000  LOCAL SECTION    0 
+013 0x00000000 0x00000000  LOCAL SECTION    0 
+014 0x00000000 0x00000000  LOCAL SECTION    0 
+015 0x00000000 0x00000000  LOCAL SECTION    0 
+016 0x00000000 0x00000000  LOCAL SECTION    0 
+017 0x00000000 0x00000000  LOCAL SECTION    0 
+018 0x00000000 0x00000000  LOCAL SECTION    0 
```

**Comment by radare** [comment]
Ping

**Comment by ret2libc** [comment]
I'm here. Changes are still the same, but I'm trying one change to make travis test together my PR with my updated tests. Let's see, if i'm successful soon that would be great, otherwise i'll just send the normal PR to r2r.

**Comment by radare** [comment]
good luck :) wish the same

**Comment by ret2libc** [comment]
See https://github.com/radare/radare2-regressions/pull/1212

File: libr/core/cbin.c [hunk 1]
```diff
@@ -1235,6 +1235,50 @@ static void set_bin_relocs(RCore *r, RBinReloc *reloc, ut64 addr, Sdb **db, char
...
+	cdsz = info? (info->bits == 64? 8: info->bits == 32? 4: info->bits == 16 ? 4: 0): 0;
+	if (cdsz == 0)
+		return;
+
+	section = r_io_section_vget (r->io, addr);
+	if (!section || section->flags & R_IO_EXEC)
+		return;
+
+	if (IS_MODE_SET(mode)) {
+		r_meta_add(r->anal, R_META_TYPE_DATA, reloc->vaddr, reloc->vaddr + cdsz, NULL);
```
Comment (radare) [inline libr/core/cbin.c]: space before (

File: libr/core/cbin.c [hunk 2]
```diff
@@ -1235,6 +1235,50 @@ static void set_bin_relocs(RCore *r, RBinReloc *reloc, ut64 addr, Sdb **db, char
...
+/* Define new data at relocation address if it's not in an executable section */
+static void add_metadata(RCore *r, RBinReloc *reloc, ut64 addr, int mode) {
+	RBinFile * binfile = r->bin->cur;
+	RBinObject *binobj = binfile ? binfile->o: NULL;
+	RBinInfo *info = binobj ? binobj->info: NULL;
+	RIOSection *section;
+	int cdsz;
+
+	cdsz = info? (info->bits == 64? 8: info->bits == 32? 4: info->bits == 16 ? 4: 0): 0;
+	if (cdsz == 0)
```
Comment (radare) [inline libr/core/cbin.c]: braces

File: libr/core/cbin.c [hunk 3]
```diff
@@ -1235,6 +1235,50 @@ static void set_bin_relocs(RCore *r, RBinReloc *reloc, ut64 addr, Sdb **db, char
...
+		return;
+
+	section = r_io_section_vget (r->io, addr);
+	if (!section || section->flags & R_IO_EXEC)
```
Comment (radare) [inline libr/core/cbin.c]: braces

File: libr/core/cbin.c [hunk 4]
```diff
@@ -1235,6 +1235,50 @@ static void set_bin_relocs(RCore *r, RBinReloc *reloc, ut64 addr, Sdb **db, char
...
+static bool is_section_symbol(RBinSymbol *s) {
+	/* workaround for some bin plugs (e.g. ELF) */
+	if (!s || *s->name) {
+		return false;
+	}
+	return (s->type && !strcmp (s->type, "SECTION"));
+}
+
+static bool is_section_reloc(RBinReloc *r) {
+	return is_section_symbol(r->symbol);
```
Comment (radare) [inline libr/core/cbin.c]: spaces before (

File: libr/core/cbin.c [hunk 5]
```diff
@@ -1288,9 +1331,7 @@ static int bin_relocs(RCore *r, int mode, int va) {
 				r_cons_printf ("f %s%s%s @ 0x%08"PFMT64x"\n",
 					r->bin->prefix ? r->bin->prefix : "reloc.",
 					r->bin->prefix ? "." : "", name, addr);
-				if (cdsz) {
-					r_cons_printf ("f Cd %d @ 0x%08"PFMT64x"\n", cdsz, addr);
-				}
+				add_metadata(r, reloc, addr, mode);
```
Comment (radare) [inline libr/core/cbin.c]: space

**Comment by MaskRay** [comment]
For `bin/p/bin_elf.c:_patch_reloc`, LLD's unified relocation (arch-dependent -> arch-agnostic) approach seems good https://github.com/llvm-mirror/lld/blob/master/ELF/Arch/X86_64.cpp#L104

**Pull request merged**